docs/ProgrammingConcerns.md

```diff
@@ -0,0 +1,170 @@
+Use `FHU::Filesystem::CopyFile` instead.
+
+### `std::fstream`
+This API always allocates memory and should be avoided.
+Use a combination of open and fextl::string APIs instead of fstream.
+
+### `std::string`
+Use `fextl::string` instead.
+
+#### std::to_string
+Use `fextl::fmt::format` instead.
+
+### APIs that FEX doesn't have a replacement for
+Don't use any of these APIs in FEXLoader/FEXInterpreter. Shoutout to
+[this](https://stackoverflow.com/questions/43056338/standard-library-facilities-which-allocate-but-dont-use-an-allocator) StackOverflow post for this
+huge list.
+
+#### `std::any`
+#### `std::function`
+Ensure the std::function doesn't allocate memory.
+#### `std::valarray`
+#### `std::filebuf`
+#### `std::inplace_merge`
+#### `<stdexcept>`
+#### `std::boyer_moore_searcher`
+#### `std::filesystem::path`
+#### `std::filesystem::directory_iterator`
+#### `std::regex`
+#### `std::thread`
+#### `std::async`
+#### `std::packaged_task`
+#### `std::promise`
+#### `<iostream>`
+#### Remember this is not an all-encompassing list! We may find APIs that still allocate memory and need to be avoided!
+
+### Regular memory allocation routines.
+Don't use these directly as they will the glibc allocator.
+
+#### mmap
+Use `FEXCore::Allocator::mmap`
+
+#### munmap
+Use `FEXCore::Allocator::munmap`
+
+#### malloc
+Use `FEXCore::Allocator::malloc`
+
+#### calloc
+Use `FEXCore::Allocator::calloc`
+
+#### memalign 
+Use `FEXCore::Allocator::memalign`
+
+#### valloc
+Use `FEXCore::Allocator::valloc`
+
+#### posix_memalign
+Use `FEXCore::Allocator::posix_memalign`
+
+#### realloc
+Use `FEXCore::Allocator::realloc`
+
+#### free
+Use `FEXCore::Allocator::free`
+
+#### aligned_alloc
+Use `FEXCore::Allocator::aligned_alloc`
+
+#### __libc_malloc
+**Don't use**
+
+#### __libc_calloc
+**Don't use**
+
+#### __libc_memalign
+**Don't use**
+
+#### __libc_valloc
+**Don't use**
+
+#### __posix_memalign
+**Don't use**
+
+#### __malloc_usable_size
+**Don't use**
+
+## FEX thunk considerations
+Now that we understand the problems of FEX memory allocations. How do these problems affect FEX's thunks?
+
+### Don't use FEXCore APIs that allocate memory through FEX's allocator
+It's as simple as that. Thunks need to use the glibc allocator so their memory allocations end up using the host glibc allocator.
+
+If there becomes a point in which FEX's thunks need to allocate memory outside of the 32-bit virtual address space then FEX can expose APIs for this.
+Currently this isn't supported but it can be as necessary.
+
+## How does FEX ensure that this paradigm doesn't break?
+FEX has the cmake option `ENABLE_GLIBC_ALLOCATOR_HOOK_FAULT` to hook in to glibc's allocator and fault if anything is allocating through it.
+This can't be used with thunks for thunk testing as those actually use the glibc allocator.
+With this option enabled, FEX's CI will run as many tests as possible (None with thunks) to ensure that FEX isn't allocating memory through glibc.
```

Comment:
`as many tests as possible (None with thunks)` seems confusing here. Is "no thunks" the only restriction? If yes, this is already implied by the previous sentence. If no, the actual restrictions should be elaborated. If no and the restrictions aren't all that important, I think it's fine to say "CI will run FEX's test suite with extra verification to ensure that FEX isn't allocating memory through glibc".

Reply:
Thinking further about it, a better approach is probably to rewire this and the previous sentence into something like:
`CI will run FEX's test suite with extra verification to ensure FEX makes no allocations are made through glibc. Thunking must be disabled for this run, since thunks are by design the only place where glibc allocation still happen.`

Reply:
Modified